syntax="proto2";

package npctransport;

message Configuration {
  message IntRange {
    required int32 lower=1;
    optional int32 upper=2;
    optional int32 steps=3;
  };
  message FloatRange {
    required double lower=1;
    optional double upper=2;
    optional int32 steps=3;
    optional double base=4;
  };
  message FGConfiguration {
    required IntRange number_of_beads=1;
    required IntRange number=2;
    required FloatRange radius=3;
    required IntRange interactions=4;
    optional FloatRange rest_length_factor=5;
    optional FloatRange D_factor=6;
  };
  message FloatConfiguration {
    required IntRange number=1;
    required FloatRange radius=2;
    required IntRange interactions=3;
    optional FloatRange D_factor=4;
  };
  message Interaction {
    required string type0=1;
    required string type1=2;
    required IntRange on_or_off=3;
    optional FloatRange nonspecific_attr_range=4;
    optional FloatRange nonspecific_attr_k=5;
    optional FloatRange site_attr_range=6;
    optional FloatRange site_attr_k=7;
  };
  message Slab {
    required IntRange on_or_off=1;
    required FloatRange height=2;
    required FloatRange radius=3;
    optional FloatRange width=4;
  }
  message Box {
    required IntRange on_or_off=1;
    required FloatRange size=2;
  };
  optional FloatRange interaction_k=3;
  optional FloatRange backbone_spring_k=4;
  optional FloatRange time_step_factor=5;
  optional Box box=6;
  optional Slab slab=18;
  optional FloatRange slack=7;
  optional IntRange number_of_trials=8;
  optional IntRange number_of_frames=11;
  repeated FGConfiguration fgs=9;
  repeated FloatConfiguration floaters=10;
  repeated Interaction interactions=14;
  optional int32 dump_interval=12;
  optional FloatRange nonspecific_range=13;
  optional FloatRange nonspecific_k=16;
  optional FloatRange angular_D_factor=15;
  optional int32 statistics_interval=17;
  optional FloatRange default_repulsive_range=19;
  optional FloatRange default_repulsive_k=20;
  optional FloatRange range=21;
}

message Assignment {
  message IntAssignment {
    required int32 value=1;
    optional int32 index=2;
  };
  message FloatAssignment {
    required double value=1;
    optional int32 index=2;
  };
  message FGAssignment {
    required IntAssignment number_of_beads=1;
    required IntAssignment number=2;
    required FloatAssignment radius=3;
    required IntAssignment interactions=4;
    optional FloatAssignment rest_length_factor=5;
    optional FloatAssignment D_factor=6;
  };
  message FloaterAssignment {
    required IntAssignment number=1;
    required FloatAssignment radius=2;
    required IntAssignment interactions=3;
    optional FloatAssignment D_factor=4;
  };
  message InteractionAssignment {
    required string type0=1;
    required string type1=2;
    required IntAssignment on_or_off=3;
    optional FloatAssignment nonspecific_attr_range=4;
    optional FloatAssignment nonspecific_attr_k=5;
    optional FloatAssignment site_attr_range=6;
    optional FloatAssignment site_attr_k=7;
  };
  message Slab {
    required IntAssignment on_or_off=1;
    required FloatAssignment height=2;
    required FloatAssignment radius=3;
    optional FloatAssignment width=4;
  };
  message Box {
    required IntAssignment on_or_off=1;
    required FloatAssignment size=2;
  };
  optional int32 work_unit=1;
  optional IntAssignment number_of_work_units=2;
  optional FloatAssignment interaction_k=3;
  optional FloatAssignment backbone_spring_k=4;
  optional FloatAssignment time_step_factor=5;
  optional Box box=6;
  optional Slab slab=21;
  optional FloatAssignment slack=7;
  optional IntAssignment number_of_trials=8;
  optional IntAssignment number_of_frames=12;
  repeated FGAssignment fgs=9;
  repeated FloaterAssignment floaters=10;
  repeated IntAssignment grid_index=11;
  optional int32 dump_interval=13;
  optional FloatAssignment nonspecific_range=14;
  optional FloatAssignment nonspecific_k=19;
  repeated InteractionAssignment interactions=15;
  optional double time_step=16;
  optional FloatAssignment angular_D_factor=18;
  optional int32 statistics_interval=20;
  optional FloatAssignment default_repulsive_range=22;
  optional FloatAssignment default_repulsive_k=23;
  optional FloatAssignment range=24;
}

message Statistics {
  message FGStats {
    optional double volume=1 [default=0];
    optional double length=2 [default=0];
    optional double radius_of_gyration=3 [default=0];
    optional double particle_diffusion_coefficient=4 [default=0];
    optional double particle_correlation_time=5 [default=0];
    optional double chain_diffusion_coefficient=6 [default=0];
    optional double chain_correlation_time=7 [default=0];
    optional double local_diffusion_coefficient=8 [default=0];
  }
  message FloatStats {
    optional double interactions=1 [default=0];
    optional double interacting=2 [default=0];
    optional double interaction_partner_chains=3 [default=0];
    optional double interaction_partner_beads=4 [default=0];
    optional double diffusion_coefficient=5 [default=0];
    optional double correlation_time=6 [default=0];
  }
  message InteractionStats {
    required string type0=1;
    required string type1=2;
    optional double avg_contacts_per_particle0=3;
    optional double avg_contacts_per_particle1=4;
    optional double avg_pct_bound_particles0=5;
    optional double avg_pct_bound_particles1=6;
  }
  repeated FGStats fgs=1;
  repeated FloatStats floaters=2;
  repeated InteractionStats interactions=6;
  optional double energy_per_particle=3 [default=0];
  optional int32 number_of_frames=4 [default=0];
  optional double seconds_per_iteration=5 [default=0];
}
